Use correct socket option level and name to get/set freebind flag

jsitnicki opened the pull request:
`freebind()`/`set_freebind()` are currently getting/setting the reuseport flag because we are passing an incorrect socket level - `SOL_SOCKET` instead of `SOL_IP`, and `SO_REUSEPORT` happens to have the same value as `IP_FREEBIND` on Linux:
```
  $ git -C ~/src/linux grep -h 'define \(SO_REUSEPORT\|IP_FREEBIND\)' -- include/uapi
  #define SO_REUSEPORT    15
  #define IP_FREEBIND     15
```
Fix it by using `(SOL_IP, IP_FREEBIND)` and `(SOL_IPV6, IPV6_FREEBIND)` socket option (level, name) to get/set the freebind flag, for INET and INET6 sockets, respectively.

Alternatively, `(SOL_IP, IP_FREEBIND)` could be used for both IPv4 and IPv6 because Linux kernel underneath toggles a flag that is shared by v4 and v6 network stack. However, this is an implementation detail, and arguably not what would be expected by the user when looking at strace output.

Fixes #257.

jsitnicki commented:
Interesting, CI failures show that `libc::IPV6_FREEBIND` doesn't exist on x86_64-fuchsia.

Let me dig around. Maybe implementation needs to be split up for Linux and Fuchsia.

Thomasdezeeuw reviewed (commented):
@jsitnicki I think you're right, but man these are poorly documented... Neither https://man7.org/linux/man-pages/man7/ip.7.html or https://man7.org/linux/man-pages/man7/ipv6.7.html mentions `IPV6_FREEBIND`. I can only conclude `SOL_IPV6` is needed and that `IPV6_FREEBIND` even exists is by looking a the Linux source, at https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/net/ipv6/ipv6_sockglue.c#L638 and 
https://github.com/torvalds/linux/blob/8096acd7442e613fad0354fc8dfdb2003cceea0b/net/mptcp/sockopt.c#L456-L480.

Nice find!

But now the solution. I don't think I want the hidden call to `domain`, socket2 tries to map methods directly to system calls. Can you instead fix the current functions by changing `SOL_SOCKET` to `SOL_IP`. And add a new variant `(set)_freebind_ipv6`? It would be nice if you could add the code that checks `domain` as an example.

Thomasdezeeuw commented:
Oh, as for the missing constant on Fuschia I'll add it to libc.

Thomasdezeeuw commented:
CI blocked on https://github.com/rust-lang/libc/pull/2341.

Thomasdezeeuw commented:
@jsitnicki Fuchsia doesn't seem to support `IPV6_FREEBIND`, so you can remove support for the v6 version from it.

jsitnicki commented:
Thanks for quick feedback @Thomasdezeeuw !

> Neither https://man7.org/linux/man-pages/man7/ip.7.html or https://man7.org/linux/man-pages/man7/ipv6.7.html mentions IPV6_FREEBIND.

Right, unfortunately there are quite a few IPv6 socket options present on Linux that are not documented by the ipv6(7) man page (yet):

https://elixir.bootlin.com/linux/v5.13.10/source/include/uapi/linux/in6.h#L152

> But now the solution. I don't think I want the hidden call to domain, socket2 tries to map methods directly to system calls. 

That is fair. I was wondering how "smart" the `freebind`/`set_freebind` wrappers should be. I will split it up as suggested.

jsitnicki commented:
v2:
* fixed `freebind`/`set_freebind` to use `SOL_IP` socket level,
* added `freebind_ipv6`/`set_freebind_ipv6` to access (`SOL_IPV6`, `IPV6_FREEBIND`) socket option on Linux platforms

jsitnicki commented:
> It would be nice if you could add the code that checks domain as an example.

One moment. Forgot to add an example.

jsitnicki commented:
v3:
* added an example for `set_freebind_ipv6`

jsitnicki commented:
v4:
* fix build failure against 1.46 due to missing `std::io::ErrorKind::Unsupported` variant.

Thomasdezeeuw commented:
Thanks @jsitnicki.
